graph/src/ipfs/cache.rs

```diff
@@ -0,0 +1,286 @@
+    }
+}
+
+#[derive(Clone, CheapClone)]
+enum Cache {
+    Memory {
+        cache: Arc<Mutex<LruCache<ContentPath, Bytes>>>,
+        max_entry_size: usize,
+    },
+    Disk {
+        store: Arc<dyn ObjectStore>,
+    },
+    Redis {
+        client: Arc<RedisClient>,
+    },
+}
+
+fn log_object_store_err(logger: &Logger, e: &object_store::Error, log_not_found: bool) {
+    if log_not_found || !matches!(e, object_store::Error::NotFound { .. }) {
+        warn!(
+            logger,
+            "Failed to get IPFS object from disk cache; fetching from IPFS";
+            "error" => e.to_string(),
+        );
+    }
+}
+
+fn log_redis_err(logger: &Logger, e: &IpfsError) {
+    warn!(
+        logger,
+        "Failed to get IPFS object from Redis cache; fetching from IPFS";
+        "error" => e.to_string(),
+    );
+}
+
+impl Cache {
+    async fn new(
+        logger: &Logger,
+        capacity: usize,
+        max_entry_size: usize,
+        path: Option<PathBuf>,
+    ) -> IpfsResult<Self> {
+        match path {
+            Some(path) if path.starts_with("redis://") => {
+                let path = path.to_string_lossy();
+                let client = RedisClient::new(logger, path.as_ref())
+                    .await
+                    .map(Arc::new)
+                    .map_err(|e| IpfsError::InvalidCacheConfig {
+                        source: anyhow!("Failed to create IPFS Redis cache at {path}: {e}"),
+                    })?;
+                Ok(Cache::Redis { client })
+            }
+            Some(path) => {
+                let fs = LocalFileSystem::new_with_prefix(&path).map_err(|e| {
+                    IpfsError::InvalidCacheConfig {
+                        source: anyhow!(
+                            "Failed to create IPFS file based cache at {}: {}",
+                            path.display(),
+                            e
+                        ),
+                    }
+                })?;
+                debug!(logger, "Using IPFS file based cache"; "path" => path.display());
+                Ok(Cache::Disk {
+                    store: Arc::new(fs),
+                })
+            }
+            None => {
+                debug!(logger, "Using IPFS in-memory cache"; "capacity" => capacity, "max_entry_size" => max_entry_size);
+                Ok(Self::Memory {
+                    cache: Arc::new(Mutex::new(LruCache::with_capacity(capacity))),
+                    max_entry_size,
+                })
+            }
+        }
+    }
+
+    async fn find(&self, logger: &Logger, path: &ContentPath) -> Option<Bytes> {
+        match self {
+            Cache::Memory {
+                cache,
+                max_entry_size: _,
+            } => cache.lock().unwrap().get(path).cloned(),
+            Cache::Disk { store } => {
+                let log_err = |e: &object_store::Error| log_object_store_err(logger, e, false);
+
+                let path = Path::from(path.cid().to_string());
```

Comment:
This is necessary to ensure that we retrieve the correct data

```suggestion
                let path = Path::from(path.to_string());
```